app/src/main/java/com/lambdasoup/quickfit/alarm/AlarmService.kt

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2016-2019 Juliane Lehmann <jl@lambdasoup.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.lambdasoup.quickfit.alarm
+
+import android.app.*
+import android.content.Context
+import android.content.Intent
+import android.media.AudioAttributes
+import android.net.Uri
+import android.os.Build
+import android.os.IBinder
+import android.os.PowerManager
+import android.provider.Settings
+import androidx.core.app.NotificationCompat
+import androidx.preference.PreferenceManager
+import com.lambdasoup.quickfit.Constants
+import com.lambdasoup.quickfit.Constants.PENDING_INTENT_WORKOUT_LIST
+import com.lambdasoup.quickfit.R
+import com.lambdasoup.quickfit.persist.QuickFitContentProvider
+import com.lambdasoup.quickfit.ui.WorkoutListActivity
+import com.lambdasoup.quickfit.util.WakefulIntents
+import timber.log.Timber
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+
+/**
+ * ForegroundService, handling alarm-related background I/O work. Could also have been solved as a [androidx.core.app.JobIntentService],
+ * or as a collection of jobs for [androidx.work.WorkManager], but (a) reading the documentation says that JobScheduler is for _deferrable_
+ * I/O work, which this is not - and also no promises are given as to the timeliness of the execution of JobScheduler jobs that are free
+ * of restrictions. Although it seems to be fine in practice (on non-background crippling devices). Also (b) - just for trying it out.
+ * Currently, this project serves as an exhibition of different ways of performing background work.
+ */
+class AlarmService : Service() {
+    private val alarms by lazy { Alarms(this.applicationContext) }
```

Comment:
Not sure if `by lazy` has any benefits here, do you?

Reply:
Hmmm.

Alternatives are:
* make those two `lateinit`, do the assignments in `onCreate`.
* eagerly do the assignments on construction - but then, I'm not sure whether a `Service` is already a fully-usable object before its `onCreate` call, in particular, whether getting the `applicationContext` in this lifecycle phase is a good idea. Do you know?

Ultimately, it's a habit carried over from Views/Activities, where "just use `by lazy`" is a useful tool for such problems.

Reply:
And yes, I think that this reads better than doing the assignment in `onCreate`.

Reply:
Since the `Service` itself implements `Context` that should be available always. I mean, just try without the `lazy` and see. On though I had was that without the `lazy` you are leaving the decision on which tread this object is constructed to the framework, with `lazy` you are creating it son the main thread when you try so access it from there.